File: src/main/java/net/onelitefeather/pandorascluster/land/flag/LandFlagEntity.java
package net.onelitefeather.pandorascluster.land.flag;

import jakarta.persistence.*;
import net.onelitefeather.pandorascluster.land.Land;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Entity
public class LandFlagEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @OneToOne
    private Land land;

    @Column
    private String name;

    @Column
    private String value;

    @Column
    private byte type;

    @Enumerated(EnumType.STRING)
    private LandFlagType flagType;

    public LandFlagEntity() {
    }

    public LandFlagEntity(@NotNull Land land, @NotNull String name, @NotNull String value, byte type, @NotNull LandFlagType flagType) {
        this.land = land;
        this.name = name;
        this.value = value;
        this.type = type;
        this.flagType = flagType;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Land getLand() {
        return land;
    }

    public void setLand(Land land) {
        this.land = land;
    }

    @NotNull
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public <T> T getValue() {
        var result = switch(this.type) {
            case 0 -> this.value;
            case 1 -> Integer.getInteger(this.value);
            case 2 -> Boolean.getBoolean(this.value);
            case 3 -> Double.parseDouble(this.value);
            case 4 -> Float.parseFloat(this.value);
            case 5 -> Short.parseShort(this.value);
            case 6 -> Byte.parseByte(this.value);
            default -> null;
        };
        return (T) result;
    }

    public byte getType() {
        return type;
    }

    public void setType(byte type) {
        this.type = type;
    }

    @Nullable
    public LandFlag getFlag() {
        return LandFlag.findByName(this.name);
    }

    public void setValue(@NotNull String value) {
        this.value = value;
    }

    @NotNull
    public LandFlagType getFlagType() {
        return flagType;
    }

    public void setFlagType(@NotNull LandFlagType flagType) {
        this.flagType = flagType;
    }

    public static final class Builder {

        private String name;
        private String value;

        private Land land;
        private LandFlagType flagType;
        private byte type;

        public LandFlagEntity.Builder name(@NotNull String name) {
            this.name = name;
            return this;
        }

        public LandFlagEntity.Builder land(@NotNull Land land) {
            this.land = land;
            return this;
        }

        public LandFlagEntity.Builder withType(@NotNull LandFlagType flagType) {
            this.flagType = flagType;
            return this;
        }

        public LandFlagEntity.Builder type(byte type) {
            this.type = type;
            return this;
        }

        public LandFlagEntity.Builder value(@NotNull String value) {
            this.value = value.trim();
            return this;
        }

        public LandFlagEntity build() {
            return new LandFlagEntity(this.land, this.name, this.value, this.type, this.flagType);
        }
    }

}

File: src/main/kotlin/net/onelitefeather/pandorascluster/land/flag/LandFlagEntity.kt
package net.onelitefeather.pandorascluster.land.flag

import jakarta.persistence.*
import net.onelitefeather.pandorascluster.land.Land
import org.apache.commons.lang3.BooleanUtils
import org.apache.commons.lang3.StringUtils
import org.hibernate.Hibernate

@Entity
data class LandFlagEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,
    @Column
    val name: String? = null,
    @Column
    var value: String? = null,
    @Column
    val type: Byte? = null,
    @Column
    @Enumerated(EnumType.STRING)
    val flagType: LandFlagType? = null,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "land_id")
    val land: Land? = null
) {

    fun <T : Any> getValue(): T? {
        val result = when (this.type?.toInt()) {
            0 -> this.value
            1 -> this.value?.toIntOrNull()
            2 -> this.value?.toBooleanStrictOrNull()
            3 -> this.value?.toDoubleOrNull()
            4 -> this.value?.toFloatOrNull()
            5 -> this.value?.toShortOrNull()
            6 -> this.value?.toByteOrNull()
            else -> null
        }
        return result as T?
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || Hibernate.getClass(this) != Hibernate.getClass(other)) return false
        other as LandFlagEntity

        return id != null && id == other.id
    }

    override fun hashCode(): Int = javaClass.hashCode()

    @Override
    override fun toString(): String {
        return this::class.simpleName + "(id = $id , name = $name , value = $value , type = $type , flagType = $flagType )"
    }
}

fun isValidValue(landFlag: LandFlag, value: String): Boolean {
    return when(landFlag.type.toInt()) {
        0 -> !StringUtils.isNumeric(value) && !BooleanUtils.isNotFalse(value.toBoolean()) && !BooleanUtils.isNotTrue(value.toBoolean())
        1 -> value.toIntOrNull() != null
        2 -> value.toBooleanStrictOrNull() != null
        3 -> value.toDoubleOrNull() != null
        4 -> value.toFloatOrNull() != null
        5 -> value.toShortOrNull() != null
        6 -> value.toByteOrNull() != null
        else -> false
    }
}

fun getDefaultFlag(landFlag: LandFlag) : LandFlagEntity = LandFlagEntity(
    landFlag.ordinal.toLong(), landFlag.name,
    landFlag.defaultValue.toString(), landFlag.type,
    landFlag.landFlagType,
    null)
